src/append/rolling_file/policy/compound/roll/fixed_window.rs

```diff
@@ -94,57 +100,125 @@ impl FixedWindowRoller {
 }
 
 impl Roll for FixedWindowRoller {
+    #[cfg(not(feature = "async_rotation"))]
     fn roll(&self, file: &Path) -> Result<(), Box<dyn Error + Sync + Send>> {
         if self.count == 0 {
             return fs::remove_file(file).map_err(Into::into);
         }
 
-        let dst_0 = self.pattern.replace("{}", &self.base.to_string());
+        rotate(
+            self.pattern.clone(),
+            self.compression.clone(),
+            self.base,
+            self.count,
+            file.to_path_buf(),
+        )?;
+
+        Ok(())
+    }
 
-        if let Some(parent) = Path::new(&dst_0).parent() {
-            fs::create_dir_all(parent)?;
+    #[cfg(feature = "async_rotation")]
+    fn roll(&self, file: &Path) -> Result<(), Box<dyn Error + Sync + Send>> {
+        if self.count == 0 {
+            return fs::remove_file(file).map_err(Into::into);
         }
 
-        // In the common case, all of the archived files will be in the same
-        // directory, so avoid extra filesystem calls in that case.
-        let parent_varies = match (
-            Path::new(&dst_0).parent(),
-            Path::new(&self.pattern).parent(),
-        ) {
-            (Some(a), Some(b)) => a != b,
-            _ => false, // Only case that can actually happen is (None, None)
-        };
+        // rename the file
+        let temp = make_temp_file_name(file);
+        move_file(file, &temp)?;
 
-        for i in (self.base..self.base + self.count - 1).rev() {
-            let src = self.pattern.replace("{}", &i.to_string());
-            let dst = self.pattern.replace("{}", &(i + 1).to_string());
+        {
+            // wait for the previous call to end
+            let _lock = self.lock.lock();
```

Comment:
I'm curious, is this lock call necessary? The comment states that we're waiting for the previous call to end, but it's not clear what the previous call is. Neither `make_temp_file_name` or `move_file` acquire the lock. `roll` itself, appears to be within the scope of the write lock in `append` in the impl of `Append` for `RollingFileAppender`, so from that perspective, I'm not sure that a concurrent call for `roll` occurs there.